### PR TITLE
Extend RSAlgorithmFactory to support all ctors of RS256Algorithm

### PR DESCRIPTION
**Pull request by flerka**
This is my draft implantation for https://github.com/jwt-dotnet/jwt/issues/220

### CODE REVIEW

**Comment by abatishchev** [comment]
Hi, thanks for the contribution!
Can you explain the idea behind heaving `Func` over instantiating an algorithm directly?

**Comment by flerka** [comment]
This was done to avoid confusing if inside switch statement. All Func can be rewritten with:

```csharp
            switch (algorithm)
            {
                case JwtHashAlgorithm.RS256:
                    {
                        if (_publicKey != null)
                        {
                            return new RS256Algorithm(publicKey);
                        }

                        var certificate = _certFactory();
#if NETSTANDARD1_3
            return new RS256Algorithm((RSACryptoServiceProvider)certificate.GetRSAPublicKey(), certificate.GetRSAPrivateKey());
#else
                        return new RS256Algorithm((RSACryptoServiceProvider)certificate.PublicKey.Key, (RSA)certificate.PrivateKey);
#endif

                    }
                default:
                    throw new NotSupportedException($"For algorithm {Enum.GetName(typeof(JwtHashAlgorithm), algorithm)} please use the appropriate factory by implementing {nameof(IAlgorithmFactory)}");
            }
```
If this is better from your point of view, I can rollback previous changes and commit this. What should I do @abatishchev ?

**Comment by abatishchev** [comment]
I was thinking more about this factory and realized we might not need it at all. Since it supports a single algorithm anyway. So it just mimics the algorithm's own constrictors. Does it add any value in top of it?
The initial ideas was to support multiple algorithms, help to instantiate one or another. But that has never happened.
So I'd remove it together. What do you think?

**Comment by abatishchev** [comment]
I'd remove `JwtHashAlgorithm` as well. Wondering what's the percentage of the users of this library uses symmetric algorithms still. Hope that not much.

**Comment by abatishchev** [comment]
Actually I got your idea and I like it. It's better than mine in #223. Can you please rebase on the latest master?

File: src/JWT/Algorithms/RSAlgorithmFactory.cs [hunk 1]
```diff
@@ -8,31 +8,48 @@ namespace JWT.Algorithms
...
+        {
             _certFactory = certFactory;
+            _createAlgorithmMethod = CreateWithCertificate;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RSAlgorithmFactory"/> class
+        /// </summary>
+        /// <param name="publicKey">The RSA service provider which will be used to instantiate <see cref="RS256Algorithm" /></param>
+        public RSAlgorithmFactory(RSACryptoServiceProvider publicKey) =>
```
Comment (abatishchev) [inline src/JWT/Algorithms/RSAlgorithmFactory.cs]: Let's accept just `RSA` here

File: src/JWT/Algorithms/RSAlgorithmFactory.cs [hunk 2]
```diff
@@ -8,31 +8,48 @@ namespace JWT.Algorithms
...
     {
         private readonly Func<X509Certificate2> _certFactory;
+        private readonly Func<RS256Algorithm> _createAlgorithmMethod;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="RSAlgorithmFactory"/> class
         /// </summary>
         /// <param name="certFactory">Func that returns <see cref="X509Certificate2" /> which will be used to instantiate <see cref="RS256Algorithm" /></param>
-        public RSAlgorithmFactory(Func<X509Certificate2> certFactory) =>
+        public RSAlgorithmFactory(Func<X509Certificate2> certFactory)
```
Comment (abatishchev) [inline src/JWT/Algorithms/RSAlgorithmFactory.cs]: Will also need a ctor that accepts both `RSA publicKey` and `RSA privateKey`

**Comment by abatishchev** [comment]
@flerka thanks for your contribution, much appreciated.

**Pull request merged**